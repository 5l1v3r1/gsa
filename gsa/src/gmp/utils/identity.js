/* Greenbone Security Assistant
 *
 * Authors:
 * Bj√∂rn Ricks <bjoern.ricks@greenbone.net>
 * Steffen Waterkamp <steffen.waterkamp@greenbone.net>
 *
 * Copyright:
 * Copyright (C) 2016 - 2018 Greenbone Networks GmbH
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
export const {isArray} = global.Array;

export function is_defined(value) {
  return value !== undefined;
}

export function has_value(value) {
  return value !== null && is_defined(value);
}

export function isObject(value) {
  return value !== null && typeof value === 'object';
}

export function isString(value) {
  return typeof value === 'string';
}

export function isNumber(value) {
  return typeof value === 'number';
}

export function isFunction(value) {
  return typeof value === 'function';
}

export function isJsDate(value) {
  return Object.prototype.toString.call(value) === '[object Date]';
}

export const isModelElement = elem => is_defined(elem) &&
  isString(elem._id) && elem._id.length > 0;

// vim: set ts=2 sw=2 tw=80:
