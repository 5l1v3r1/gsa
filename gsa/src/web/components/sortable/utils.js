/* Greenbone Security Assistant
 *
 * Authors:
 * Bj√∂rn Ricks <bjoern.ricks@greenbone.net>
 *
 * Copyright:
 * Copyright (C) 2018 Greenbone Networks GmbH
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
import uuid from 'uuid/v4';

import {createRow} from 'gmp/commands/dashboards';

export const createItem = (props, uuidFunc = uuid) => {
  const id = uuidFunc();

  return {
    id,
    ...props,
  };
};

export const removeItem = (rows, itemId) => rows.map(row => ({
  ...row,
  items: row.items.filter(item => item.id !== itemId),
})).filter(row => row.items.length > 0);

export const updateRow = (row, data) => ({
  ...row,
  ...data,
});

export const convertDefaultContent = (defaultContent = [], uuidFunc = uuid) => {
  return {
    rows: defaultContent.map(row => createRow(
      row.map(item => createItem({name: item}, uuidFunc)), undefined, uuidFunc
    )),
  };
};

// vim: set ts=2 sw=2 tw=80:
