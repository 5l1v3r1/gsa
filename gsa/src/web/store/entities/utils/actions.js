/* Greenbone Security Assistant
 *
 * Authors:
 * Bj√∂rn Ricks <bjoern.ricks@greenbone.net>
 *
 * Copyright:
 * Copyright (C) 2018 Greenbone Networks GmbH
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */

export const types = {
  ENTITIES_LOADING_REQUEST: 'ENTITIES_LOADING_REQUEST',
  ENTITIES_LOADING_SUCCESS: 'ENTITIES_LOADING_SUCCESS',
  ENTITIES_LOADING_ERROR: 'ENTITIES_LOADING_ERROR',
};

export const createEntitiesActionCreators = entityType => ({
  request: filter => ({
    type: types.ENTITIES_LOADING_REQUEST,
    entityType,
    filter,
  }),
  success: (data, filter) => ({
    type: types.ENTITIES_LOADING_SUCCESS,
    entityType,
    filter,
    data,
  }),
  error: (error, filter) => ({
    type: types.ENTITIES_LOADING_ERROR,
    entityType,
    filter,
    error,
  }),
});

export const createLoadAllFunc = ({
  selector,
  actionCreators,
  entityType,
}) => ({gmp, filter, ...props}) => (dispatch, getState) => {
    const rootState = getState();
    const state = selector(rootState);

    if (state.isLoadingEntities(filter)) {
      // we are already loading data
      return Promise.resolve();
    }

    dispatch(actionCreators.request(filter));

    return gmp[entityType].getAll({filter}).then(
      response => dispatch(actionCreators.success(response.data, filter)),
      error => dispatch(actionCreators.error(error, filter)),
    );
  };

// vim: set ts=2 sw=2 tw=80:
