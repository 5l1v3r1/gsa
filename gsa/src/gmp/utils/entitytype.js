/* Greenbone Security Assistant
 *
 * Authors:
 * Bj√∂rn Ricks <bjoern.ricks@greenbone.net>
 *
 * Copyright:
 * Copyright (C) 2018 Greenbone Networks GmbH
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
import {is_defined} from './identity';

export const getEntityType = (model = {}) => {
  const {entity_type: entityType} = model;
  if (entityType === 'info' && is_defined(model.info_type)) {
    return model.info_type;
  }
  if (entityType === 'asset' && is_defined(model.asset_type)) {
    return model.asset_type;
  }
  return entityType;
};

export const pluralizeType = type => type[type.length - 1] === 's' ||
  type === 'info' || type === 'version' ? type : type + 's';

const types = {
  config: 'scanconfig',
  cert_bund_adv: 'certbund',
  dfn_cert_adv: 'dfncert',
  os: 'operatingsystem',
  port_list: 'portlist',
  report_format: 'reportformat',
};

export const normalizeType = type => {
  const ctype = types[type];
  return is_defined(ctype) ? ctype : type;
};

// vim: set ts=2 sw=2 tw=80:
