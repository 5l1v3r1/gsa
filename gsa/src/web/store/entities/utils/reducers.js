/* Greenbone Security Assistant
 *
 * Authors:
 * Bj√∂rn Ricks <bjoern.ricks@greenbone.net>
 *
 * Copyright:
 * Copyright (C) 2018 Greenbone Networks GmbH
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
import {combineReducers} from 'redux';

import {is_defined} from 'gmp/utils/identity';

export const filterIdentifier = filter => is_defined(filter) ?
  `filter:${filter.toFilterString()}` :
  'default';

export const createReducer = types => {

  const isLoading = (state = false, action) => {
    switch (action.type) {
      case types.REQUEST:
        return true;
      case types.SUCCESS:
      case types.ERROR:
        return false;
      default:
        return state;
    }
  };

  const error = (state = null, action) => {
    switch (action.type) {
      case types.SUCCESS:
        return null;
      case types.ERROR:
        return action.error;
      default:
        return state;
    }
  };

  const entities = (state = [], action) => {
    switch (action.type) {
      case types.SUCCESS:
        const {data = []} = action;
        return data.map(entity => entity.id);
      default:
        return state;
    }
  };

  const byId = (state = {}, action) => {
    switch (action.type) {
      case types.SUCCESS:
        const {data = []} = action;
        const nextState = {
          ...state,
        };
        data.forEach(d => nextState[d.id] = d);
        return nextState;
      default:
        return state;
    }
  };

  const combinedReducer = combineReducers({
    isLoading,
    error,
    entities,
  });

  return (state = {}, action) => {
    switch (action.type) {
      case types.REQUEST:
      case types.SUCCESS:
      case types.ERROR:
        const filterString = filterIdentifier(action.filter);
        return {
          ...state,
          byId: byId(state.byId, action),
          [filterString]: combinedReducer(state[filterString], action),
        };
      default:
        return state;
    }
  };
};

// vim: set ts=2 sw=2 tw=80:
