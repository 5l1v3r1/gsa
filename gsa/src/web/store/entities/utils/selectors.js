/* Greenbone Security Assistant
 *
 * Authors:
 * Bj√∂rn Ricks <bjoern.ricks@greenbone.net>
 *
 * Copyright:
 * Copyright (C) 2018 Greenbone Networks GmbH
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
import {is_defined, has_value} from 'gmp/utils/identity';

class EntitiesSelector {

  constructor(state = {}) {
    this.state = state;
  }

  _getByFilter(filter) {
    const filterString = is_defined(filter) ? filter.toFilterString() :
      'default';
    return this.state[filterString];
  }

  getIsLoading(filter) {
    const state = this._getByFilter(filter);
    return is_defined(state) ? state.isLoading : false;
  }

  getError(filter) {
    const state = this._getByFilter(filter);
    return is_defined(state) ? state.error : undefined;
  }

  getEntities(filter) {
    const state = this._getByFilter(filter);
    if (is_defined(state) && has_value(state.entities)) {
      return state.entities.map(id => this.getEntity(id)).filter(is_defined);
    }
    return undefined;
  }

  getEntity(id) {
    return this.state.byId[id];
  }
};

export const createSelector = name => rootState =>
  new EntitiesSelector(rootState.entities[name]);

// vim: set ts=2 sw=2 tw=80:
