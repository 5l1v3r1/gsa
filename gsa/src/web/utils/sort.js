/* Greenbone Security Assistant
 *
 * Authors:
 * Timo Pollmeier <timo.pollmeier@greenbone.net>
 * Bj√∂rn Ricks <bjoern.ricks@greenbone.net>
 *
 * Copyright:
 * Copyright (C) 2017 - 2018 Greenbone Networks GmbH
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
import logger from 'gmp/log';

import {parseInt, parseFloat} from 'gmp/parser';

import {isDefined, isFunction} from 'gmp/utils/identity';

const log = logger.getLogger('web.utils.sort');

const genericCompareAsc = (a, b) => {
  if (a > b) {
    return +1;
  }
  else if (b > a) {
    return -1;
  }
  return 0;
};

const genericCompareDesc = (a, b) => {
  if (a < b) {
    return +1;
  }
  else if (b < a) {
    return -1;
  }
  return 0;
};

const getProperty = (object, property) => {
  try {
    if (isFunction(property)) {
      return property(object);
    }

    return object[property];
  }
  catch (err) {
    log.error('Could not get property', property, object);
    return undefined;
  }
};

export const ipToNumber = original => {
  if (!isDefined(original)) {
    return undefined;
  }

  const split = original.split('.');
  if (split.length === 4) { // should be an ipv4 address
    let ret = 0;
    for (const item of split) {
      ret = ret << 8; // eslint-disable-line no-bitwise
      const number = parseInt(item);

      if (!isDefined(number)) { // wasn't a number. it's not an ip
        return original;
      }

      ret = ret | number; // eslint-disable-line no-bitwise
    }
    return ret;
  }

  // TODO support ipv6

  return original; // use original value for comparison
};

const getValue = (convertFunc, value, property, undefinedVal) => {
  const val = convertFunc(getProperty(value, property));

  return isDefined(val) ? val : undefinedVal;
};

const makeCompare = convertFunc => (property, undefinedVal) => reverse => {
  const valCompare = reverse ? genericCompareDesc : genericCompareAsc;

  return (a, b) => valCompare(
    getValue(convertFunc, a, property, undefinedVal),
    getValue(convertFunc, b, property, undefinedVal),
  );
};

export const makeCompareString = makeCompare(value => '' + value);

export const makeCompareNumber = makeCompare(parseFloat);

export const makeCompareDate = makeCompare(value => value);

export const makeCompareIp = makeCompare(ipToNumber);

export const makeCompareSeverity = (name = 'severity') =>
  makeCompareNumber(name, 0);

// vim: set ts=2 sw=2 tw=80:
