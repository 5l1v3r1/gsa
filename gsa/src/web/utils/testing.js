/* Copyright (C) 2018 Greenbone Networks GmbH
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
import 'jest-styled-components';

import React from 'react';

import {render as reactTestingRender} from 'react-testing-library';

import {Router} from 'react-router-dom';

import {Provider} from 'react-redux';

import {createMemoryHistory} from 'history';

import {
  toHaveAttribute,
  toHaveTextContent,
} from 'jest-dom';

import {hasValue} from 'gmp/utils/identity';

import configureStore from 'web/store';

expect.extend({
  toHaveAttribute,
  toHaveTextContent,
});

export * from 'react-testing-library';

export const render = ui => {
  const {container, ...other} = reactTestingRender(ui);
  return {
    container,
    element: hasValue(container) ? container.firstChild : undefined,
    ...other,
  };
};

export const rendererWithRouter = (
  history = createMemoryHistory({initialEntries: ['/']}),
) => ({
  render: ui => render(
    <Router history={history}>{ui}</Router>
  ),
  history,
});

export const rendererWithStore = (store = configureStore()) => ({
  render: ui => render(
    <Provider store={store}>{ui}</Provider>
  ),
  store,
});

export const rendererWithStoreAndRouter = (
  store = configureStore(),
  history = createMemoryHistory({initialEntries: ['/']}),
) => ({
  store,
  history,
  render: ui => render(
    <Provider store={store}>
      <Router history={history}>
        {ui}
      </Router>
    </Provider>
  ),
});

// vim: set ts=2 sw=2 tw=80:
