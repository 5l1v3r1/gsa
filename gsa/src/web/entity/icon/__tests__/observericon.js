/* Copyright (C) 2019 Greenbone Networks GmbH
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 */
import React from 'react';

import Task from 'gmp/models/task';
import Audit from 'gmp/models/audit';
import Target from 'gmp/models/target';

import {render} from 'web/utils/testing';

import ObserverIcon from '../observericon';

describe('Entity ObserverIcon component tests', () => {
  test('should render task if the owner is not the current user', () => {
    const task = Task.fromObject({owner: 'foo'});

    const {element} = render(<ObserverIcon entity={task} userName={'bar'} />);

    expect(element).toBeInTheDocument();
  });

  test('should render non-task if the owner is not the current user', () => {
    const audit = Audit.fromElement({owner: {name: 'foo'}});

    const {element} = render(<ObserverIcon entity={audit} userName={'bar'} />);

    expect(element).toBeInTheDocument();
  });

  test('should not render task if the owner is the current user', () => {
    const task = Task.fromObject({owner: 'foo'});

    const {element} = render(<ObserverIcon entity={task} userName={'foo'} />);

    expect(element).toEqual(null);
  });

  test('should not render non-task if the owner is the current user', () => {
    const target = Target.fromElement({owner: {name: 'foo'}});

    const {element} = render(<ObserverIcon entity={target} userName={'foo'} />);

    expect(element).toEqual(null);
  });
});

// vim: set ts=2 sw=2 tw=80:
