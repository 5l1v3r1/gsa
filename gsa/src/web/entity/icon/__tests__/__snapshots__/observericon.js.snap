// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity ObserverIcon component tests should render if the owner is not the current user 1`] = `
.c1 {
  cursor: pointer;
}

.c0 {
  height: 16px;
  width: 16px;
  line-height: 16px;
}

.c0 * {
  height: inherit;
  width: inherit;
}

@media print {
  .c1 {
    display: none;
  }
}

<span
  alt="Entity owned by foo"
  class="c0 c1"
  data-testid="svg-icon"
  title="Entity owned by foo"
>
  <svg>
    view_other.svg
  </svg>
</span>
`;
